nlptest/transform/robustness.py

```diff
@@ -692,3 +691,61 @@ def search_contraction(text):
                     sample.transformations = transformations
             sample.category = "robustness"
         return sample_list
+        
+class NumberToWord(BaseRobustness):
+    alias_name = "number_to_word"
+    infEng = engine()
+
+    @staticmethod
+    def transform(sample_list: List[Sample]) -> List[Sample]:
+        """
+        Transform a list of strings to their equivalent verbal representation
+        of numbers present in the string.
+        Args:
+            sample_list: List of sentences to apply robustness.
+        Returns:
+            List of sentences that have numbers in their verbal representation.
+        """
+        for sample in sample_list:
+            results = []
+            trans = []
+            transformations = []
+            start_offset = 0
+            for token in sample.original.split():
+                if re.match(r'^\d+$', token):
```

Comment:
Maybe we can use regex directly on sample.original instead of splitting and rejoining because it has some side-effects like removing extra whitespace or replacing other whitespaces (\n \t) with just space. @Ryzxxl

Reply:
Yes, we can do it like this.
It will handles cases where the token is not entirely numeric, such as "5th" or "1234abc". It extracts the numerical digits and replaces them with their verbal representation. The remaining non-numeric characters are concatenated with the verbal representation of the digits to form the transformed token.

`class NumberToWord(BaseRobustness):
    alias_name = "number_to_word"
    infEng = engine()

    @staticmethod
    def transform(sample_list: List[Sample]) -> List[Sample]:
        """
        Transform a list of strings to their equivalent verbal representation
        of numbers present in the string.
        Args:
            sample_list: List of sentences to apply robustness.
        Returns:
            List of sentences that have numbers in their verbal representation.
        """
        for sample in sample_list:
            results = []
            trans = []
            transformations = []
            start_offset = 0
            for match in re.finditer(r'\d+', sample.original):
                token = match.group(0)
                words = NumberToWord.infEng.number_to_words(int(token), wantlist=True)
                token_len = len(token) - 1
                new_words_len = len(' '.join(words)) - 1
                trans.append(sample.original[start_offset:match.start()])
                trans.extend(words)
                transformations.append(
                    Transformation(
                        original_span=Span(start=match.start(), end=match.end(), word=token),
                        new_span=Span(start=match.start(), end=match.start() + new_words_len, word=' '.join(words)),
                        ignore=False
                    )
                )
                start_offset = match.end()
            trans.append(sample.original[start_offset:])
            results.append("".join(trans))
            sample.test_case = " ".join(results)
            sample.category = "robustness"
            sample.transformations = transformations
        return sample_list`